### ISSUE TITLE
loki-distributed gateway NGINX configuration hides ruler's API

### ISSUE THREAD

**Issue by cydergoth** [grafana/helm-charts#964]
The `ruler` component exposes these APIs

```
GET /ruler/ring
GET /loki/api/v1/rules
GET /loki/api/v1/rules/{namespace}
GET /loki/api/v1/rules/{namespace}/{groupName}
POST /loki/api/v1/rules/{namespace}
DELETE /loki/api/v1/rules/{namespace}/{groupName}
DELETE /loki/api/v1/rules/{namespace}
GET /api/prom/rules
GET /api/prom/rules/{namespace}
GET /api/prom/rules/{namespace}/{groupName}
POST /api/prom/rules/{namespace}
DELETE /api/prom/rules/{namespace}/{groupName}
DELETE /api/prom/rules/{namespace}
GET /prometheus/api/v1/rules
GET /prometheus/api/v1/alerts
```

but the NGINX rules specify :

```
location ~ /api/prom/.* {
            proxy_pass       http://{{ include "loki.queryFrontendFullname" . }}.{{ .Release.Namespace }}.svc.{{ .Values.global.clusterDomain }}:3100$request_uri;
          }
```

In distributed/microservices mode this rule misdirects `/api/prom/rules` to the `frontend` service not the `ruler` service

There is also a missing redirect for `/prometheus/api/v1/....`

Workaround: Completely override the nginx config to specify a more correct ruleset

**Comment by qdupuy** [comment]
I confirm, I have the same problem.

For loki ruler, which road did you use? @cydergoth

**Comment by cydergoth** [comment]
Good to see I'm not the only one!

Overrode the entire nginx configuration to add the missing location
directives.

I'll post my configuration when I get a chance.

**Comment by qdupuy** [comment]
@cydergoth 🖖 

Were you able to get feedback from the developers?

I tried a nginx configuration but it's not better /:

Did you find the right configuration after the API change?

**Comment by uberbrodt** [comment]
I was able to get the gateway sorted out (I think) with this in the gateway.nginxConfig value of the loki-distributed chart. This is just a snippet with the location proxies, see the helm chart values for the full file.

```
              location = / {
                return 200 'OK';
                auth_basic off;
              }

              location = /api/prom/push {
                proxy_pass       http://{{ include "loki.distributorFullname" . }}.{{ .Release.Namespace }}.svc.{{ .Values.global.clusterDomain }}:3100$request_uri;
              }

              location = /api/prom/tail {
                proxy_pass       http://{{ include "loki.querierFullname" . }}.{{ .Release.Namespace }}.svc.{{ .Values.global.clusterDomain }}:3100$request_uri;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection "upgrade";
              }

              location ~ /api/prom/alerts {
                proxy_pass       http://loki-ruler.observer.svc.cluster.local:3100$request_uri;
              }

              location ~ /api/prom/alerts/.* {
                proxy_pass       http://loki-ruler.observer.svc.cluster.local:3100$request_uri;
              }

              location ~ /api/prom/rules {
                proxy_pass       http://loki-ruler.observer.svc.cluster.local:3100$request_uri;
              }

              location ~ /api/prom/rules/.* {
                proxy_pass       http://loki-ruler.observer.svc.cluster.local:3100$request_uri;
              }

              location ~ /api/prom/.* {
                proxy_pass       http://{{ include "loki.queryFrontendFullname" . }}.{{ .Release.Namespace }}.svc.{{ .Values.global.clusterDomain }}:3100$request_uri;
              }

              location ~ /prometheus/api/.* {
                proxy_pass       http://loki-ruler.observer.svc.cluster.local:3100$request_uri;
              }

              location = /loki/api/v1/push {
                proxy_pass       http://{{ include "loki.distributorFullname" . }}.{{ .Release.Namespace }}.svc.{{ .Values.global.clusterDomain }}:3100$request_uri;
              }

              location = /loki/api/v1/rules/* {
                proxy_pass       http://loki-ruler.observer.svc.cluster.local:3100$request_uri;
              }

              location = /loki/api/v1/tail {
                proxy_pass       http://{{ include "loki.querierFullname" . }}.{{ .Release.Namespace }}.svc.{{ .Values.global.clusterDomain }}:3100$request_uri;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection "upgrade";
              }

              location ~ /loki/api/.* {
                proxy_pass       http://{{ include "loki.queryFrontendFullname" . }}.{{ .Release.Namespace }}.svc.{{ .Values.global.clusterDomain }}:3100$request_uri;
              }

              {{- with .Values.gateway.nginxConfig.serverSnippet }}
              {{ . | nindent 4 }}
              {{- end }}
            }
          }

```
My current issue now is that the S3 storage is not picking up the rules files I placed there. Going to open another issue for that.

**Comment by parkedwards** [comment]
so you can actually append the specific routes you need for the ruler by adding a yaml string to the `gateway.nginxConfig.serverSnippet` chart value, which is how i got around this problem

HOWEVER -- if you want to use the provided template helpers so that you can just interpolate the FQDN of the ruler like this:

```
  nginxConfig:
    serverSnippet: |-
      location = /api/prom/rules {
        proxy_pass       http://{{ include "loki.rulerFullname" . }}.{{ .Release.Namespace }}.svc.{{ .Values.global.clusterDomain }}:3100$request_uri;
      }
```

...because the `serverSnippet` itself is nested within a parent template, for some reason, Helm (or go templating) wont' resolve the lower variable reference.  It looks like this might be because of the way the chart itself interpolates the `serverSnippet` config value:

```
              {{- with .Values.gateway.nginxConfig.serverSnippet }}
              {{ . | nindent 4 }}
              {{- end }}
```

TLDR - 
1. think this project will need to add the ruler routes to the server config OR
2. more flexibly handle the interpolated `serverSnippet` chunk that the user defines, so that it can resolve/unfurl the `serverSnippet`'s template variables
